README.md
# VITime
CLI tool to view your timetable from terminal anytime!

<a href="https://www.python.org/"><img src="https://img.shields.io/badge/built%20with-Python3-blue.svg" alt="built with Python3" /></a>
<a href="https://badge.fury.io/py/vitime"><img src="https://img.shields.io/pypi/v/vitime?color=blue&label=pypi%20package" alt="PyPI version" height="18"></a>



## Table of contents  

- [Preview](https://github.com/Dhruv9449/VITime-CLI#preview)  
- [Installation](https://github.com/Dhruv9449/VITime-CLI#installation)
  - [PyPI](https://github.com/Dhruv9449/VITime-CLI#pypi)
  - [Source code](https://github.com/Dhruv9449/VITime-CLI#source-code)
- [Updates](https://github.com/Dhruv9449/VITime-CLI#options-and-updates)
- [Setting up](https://github.com/Dhruv9449/VITime-CLI#setting-up)
  - [Add timetable](https://github.com/Dhruv9449/VITime-CLI#vitime-timetable)
  - [Add course](https://github.com/Dhruv9449/VITime-CLI#vitime-addcourse)
- [Usage](https://github.com/Dhruv9449/VITime-CLI#usage)
  - [View today's timetable](https://github.com/Dhruv9449/VITime-CLI#vitime-today)
  - [View a day's timetable](https://github.com/Dhruv9449/VITime-CLI#vitime-showday)
  - [View full timetable](https://github.com/Dhruv9449/VITime-CLI#vitime-full)
  - [Show all courses](https://github.com/Dhruv9449/VITime-CLI#vitime-showcourses)
  - [Delete course](https://github.com/Dhruv9449/VITime-CLI#vitime-deletecourse)
  - [Delete timetable](https://github.com/Dhruv9449/VITime-CLI#vitime-deletetimetable)
- [Change log](https://github.com/Dhruv9449/VITime-CLI/blob/main/CHANGELOG.md)
- [License](https://github.com/Dhruv9449/VITime-CLI#license)
- [References](https://github.com/Dhruv9449/VITime-CLI#references)



## Preview

![VITime](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/VITime.gif)




## Installation
### PyPI
Open terminal and install the package using python package installer.
```sh
pip install vitime==0.2.1
```
**NOTE:** This might require admin access, on windows run your terminal as administrator and for linux based systems use-  
```sh
sudo pip install vitime==0.2.1
```


### Source code
- Download latest release via [zip file](https://github.com/Dhruv9449/VITime-CLI/archive/refs/tags/v0.2.1.zip) or [tar file](https://github.com/Dhruv9449/VITime-CLI/archive/refs/tags/v0.2.1.tar.gz)
- Extract the folder to your desired location
- Then install it in your desired terminal using suitable command
#### Linux or Max OS X
```sh
user@hostname:~$ sudo python3 setup.py install
```
#### Windows
Run command prompt as administrator and
```psh
C:\users\username> python setup.py install
```



## Updates

VITime has automatic update detection to notify user when a newer version is available! You can update using `vitime -u` or `vitime --update`.

![vitime -u](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_update.gif)



## Setting up
### Commands
#### `vitime addtimetable`

Command to add entire timetable.  
> **Note:** This deletes any existing timetable.  
<p>
1. Copying timetable from vtop  

![copying timetable](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/copying_timetable.gif)</p>   

<p>
2. Adding timetable using VITime  

![vitime addtimetable](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_addtimetable.gif)</p>




#### `vitime addcourse`

Command to add a course.

![vitime addcourse](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_addcourse.gif)



## Usage
### Commands
#### `vitime today`  

Shows all ongoing/left classes for the day.  
> **Note:** This does not show all the classes during the day, to view all classes use `vitable showday` command.

![vitime today](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_today.gif)





#### `vitime showday`

Shows all the classes on the day entered by the user.

![vitime showday](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_showday.gif)





#### `vitime full`

Shows the entire timetable.

![vitime full](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_full.gif)






#### `vitime showcourses`

Shows a lot of all the courses added to the timetable.

![vitime showcourses](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_showcourses.gif)





#### `vitime deletecourse`

Deletes the given course from timetable.

![vitime deletecourse](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_deletecourse.gif)




#### `vitime deletetimetable`

Deletes the entire timetable.  
> **Note:** The confirmation message is case sensitive so 'YES', 'Yes' will not work.

![vitime deletetimetable](https://github.com/Dhruv9449/VITime-CLI/blob/main/assets/vitime_deletetimetable.gif)



## License
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://github.com/Dhruv9449/VITime-CLI/blob/main/LICENSE)  

Copyright (c) 2021 Dhruv9449  
[MIT license](LICENSE)

## References
This project is inspired by [VITable](https://github.com/CartelProject/vitable) made by -  
[CartelProject](https://github.com/CartelProject)
- [Vishesh Bansal](https://github.com/VisheshBansal)
- [Gagan Malvi](https://github.com/gaganmalvi)
<br>
<br>

<p align="center">
Developed by <a href="https://github.com/Dhruv9449" target=_blank>Dhruv Shah</a>
</p>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [v0.2.1] 2021-12-05
### Added
- Auto update check! VITime will let you know if there's a newer version available!
- Update option - `vitime -u` or  `vitime --update` to update to the latest version.
- Version option :  `vitime -v` or `vitime --version` shows the version.
- Change log - To keep track of all the updates and changes and display them to the users.


VITimeCLI/functions.py
from .Slotdata import *
from datetime import datetime
from .initialize import *
import click
import re

def savecourse(Course):
    Course.type = Course.slots[0][0]
    if Course.type == 'L':
        Course.type = 'P'
    else :
        Course.type = 'L'

    if Course.type == "L":
        daysTL = daysTh
        days = list(daysTL.keys())
        times = thSlots
        allslots = theoryslots
    else:
        daysTL = daysLb
        days = list(daysTL.keys())
        times = lbSlots
        allslots = labslots

    for slot in Course.slots:
        if slot not in allslots:
            print(f"\nInvalid slots entered for {Course.code}! Please enter valid slots.\nCourse not added!")
            return False

    for slot in Course.slots:
        for day in daysTL:
            dt = {}
            dt[day]=[]
            daysl = daysTL[day]
            if slot in daysl:
                for i in range(len(daysl)):
                    if slot == daysl[i]:
                        time = times[i]
                        dt[day].append(time)
                        coursedetails = [time, Course.code, Course.name, slot, Course.type]
                        d1 = loadday(day)
                        d1.schedule.append(coursedetails)
                        d1.schedule.sort()
                        addschedule_db(d1)
    addcourse_db(Course)
    return True





@click.group(cls=CustomGroup, context_settings=help)
def vitime():
    pass

@vitime.command()
@click.pass_context
def addtimetable(ctx):
    '''
    Add your timetable and delete existing one.
    '''
    if ctx.invoke(deletetimetable):
        print("Please type 'yes' to delete existing timetable and add new one!")
        return

    print("-"*50)
    text = retrieveinput()
    if text == False:
        return
    print("\n"+"-"*50)

    data = scrape(text)
    if len(data)==0:
        print("\nInvalid input! Please try again with a valid input or use the addcourse command to add courses manually.")
        print("\nFor more detailed help on how to add timetable check -\n  https://github.com/Dhruv9449/VITime-CLI#vitime-addtimetable\n")
        return

    for i in data :
        Course = course()
        Course.code = i[0]
        Course.name = i[1]
        Course.slots = i[2]
        if Course.slots[0]=="NIL":
            continue
        if checkcourse(Course):
            continue
        if not savecourse(Course):
            continue
        print(f"{Course.code} added!")

    print('\nTimetable added!\n')



@vitime.command(cls=CustomCommand)
def showcourses():
    '''
    Shows all courses.
    '''
    print("             VITIME\n")
    cursor.execute("SELECT code, name, type, slots FROM Courses")
    courses = cursor.fetchall()
    if courses == []:
        print("No courses! Empty timetable!")
    for i in range(len(courses)):
        if courses[i][2] == 'L':
            type = '  [Theory] '
        else:
            type = '   [Lab]    '
        print(i+1, courses[i][0] + " - "+ courses[i][1])
        print(type, f"({'+'.join(eval(courses[i][3]))})")
        print()



@vitime.command(cls=CustomCommand)
def addcourse():
    '''
    Adds courses to your time table.
    '''
    print("             VITIME")
    while True:
        Course = course()
        Course.code = input("\nEnter course code : ").upper()
        if len(Course.code)<5:
            print("\nInvalid course code, please enter a valid course code.(eg - BCHY101L/CSE1002)\n")
            return
        Course.name = input("Enter course name : ")
        Course.slots = input("Please enter course slots (eg- E2+TE2) : ").upper().split("+")

        if checkcourse(Course):
            return

        check = savecourse(Course)
        if check:
            print("\nSuccessfully added course!")
        opt = input("\nDo you want to add more courses (y/n) : ").lower()[:1]
        if opt == "n":
            break
            print()



@vitime.command(cls=CustomCommand)
def deletecourse():
    '''
    Deletes a course from the timetable.
    '''
    print("             VITIME")
    cursor.execute("SELECT code, name, type, slots FROM Courses")
    courses = cursor.fetchall()
    cursor.execute("SELECT schedule FROM Schedules")
    days = cursor.fetchall()
    if len(courses) == 0 :
        print("\nYou've got no courses!")
        return
    print("Courses\n")
    for i in range(len(courses)):
        if courses[i][2] == 'L':
            type = '  [Theory] '
        else:
            type = '   [Lab]    '
        print(i+1, courses[i][0] + " - "+ courses[i][1])
        print(type, f"({'+'.join(eval(courses[i][3]))})")
        print()
    deletecourse = input(("Input enter the sno of course to be deleted from above : "))
    if not deletecourse.isdigit():
        print("\nPlease enter a valid number!")
        return
    deletecourse = int(deletecourse)
    if deletecourse>len(courses) or deletecourse<1:
        print("\nInvalid option plese enter valid option!")
        return
    courses = courses[deletecourse-1]
    coursecode, courseslot = courses[0], courses[3]
    courses = coursecode, courseslot
    cursor.execute("DELETE FROM Courses WHERE code = ? AND slots = ?", courses)
    for i in days:
        i = eval(i[0])
        k = list(i)
        for j in i:
            if j[1] == coursecode :
                k.remove(j)
        cursor.execute("UPDATE Schedules SET schedule = ? WHERE schedule = ?",
                        (str(k), str(i)))
    mycon.commit()
    print(f"\nCourse {coursecode} deleted!\n")



@vitime.command(cls=CustomCommand)
def deletetimetable():
    '''
    Deletes entire timetable including all courses.
    '''
    print("             VITIME")
    confirm = input("""\nAre you sure you want to clear all the contents of the existing timetable?
This step is irreversible. Type "yes" to confirm : """)

    if confirm != "yes" :
        print("\nTimetable not deleted, if you want to delete the timetable try again!\n")
        return True
    cursor.execute("DROP TABLE Schedules")
    cursor.execute("DROP TABLE Courses")
    initialize()
    mycon.commit()
    print("\nDeleted timetable!\n")



@vitime.command(cls=CustomCommand)
@click.option('-n', '--name', required=True, type = str,
              prompt = "Enter the day ", help = "Specifies the name of day")
def showday(name):
    '''
    Shows the timetable for the given day.
    '''
    print("             VITIME")
    Day = name.lower()
    if Day not in days :
        print("\nPlease enter a working day (eg - monday - friday)!\n")
        return
    cursor.execute("SELECT schedule FROM Schedules WHERE day = ?",(Day,))
    timetable = eval(cursor.fetchone()[0])
    print(f"\nDay's schedule : {Day}")
    if timetable == [] :
        print("No classes on this day!\n")
        return
    print()
    for i in timetable:
        print(i[1])
        print(i[2])
        print(f"Course type : {checktype(i[4])}")
        print(f"Slot : {i[3]}")
        print(f"Time : {i[0]}\n")
    print()



@vitime.command(cls=CustomCommand)
def today():
    '''
    Show's today's timetable and the classes left.
    '''
    print("             VITIME\n")
    d, time = datetime.now().strftime("%A %H:%M").lower().split()
    if d in ["saturday","sunday"] :
        print("It's holiday! Have fun!\n")
        return
    cursor.execute("SELECT schedule FROM Schedules WHERE day = ?",(d,))
    timetable = eval(cursor.fetchall()[0][0])
    ongoing = False
    for i in range(len(timetable)):
        end = timetable[i][0].split()[2]
        if end<time:
            continue
        else:
            start = timetable[i][0].split()[0]
            if start < time:
                ongoing = True
            timetable = timetable[i:]
            break
    else:
        timetable = []

    print("Today's Timetable :  Classes left\n")
    if timetable == []:
        print("No classes left to attend today!\n")
        return
    if ongoing:
        print("-> Ongoing!")
    else:
        print("-> Next class")
    for i in timetable:
        print(i[1])
        print(i[2])
        print(f"Course type : {checktype(i[4])}")
        print(f"Slot : {i[3]}")
        print(f"Time : {i[0]}\n")
    print()



@vitime.command(cls=CustomCommand)
def full():
    '''
    Shows the full timetable for all the days.
    '''
    print("             VITIME")
    cursor.execute("SELECT * FROM Schedules")
    data = cursor.fetchall()
    print("\nFull timetable")
    print("-"*50)
    for k in data:
        day, schedule = k
        schedule = eval(schedule)
        print(f"Day's schedule : {day}\n")
        for i in schedule:
            print(i[1])
            print(i[2])
            print(f"Course type : {checktype(i[4])}")
            print(f"Slot : {i[3]}")
            print(f"Time : {i[0]}\n")
        print("-"*50)


VITimeCLI/initialize.py
import os
import sqlite3 as sq
import click
import pyperclip
import re

homedir = os.path.expanduser("~")


mycon = sq.connect(homedir+"/VITime.db")
cursor = mycon.cursor()

def initialize():
    cursor.execute("""CREATE TABLE IF NOT EXISTS Schedules(day TEXT PRIMARY KEY,
                      schedule TEXT)""")
    cursor.execute("""CREATE TABLE IF NOT EXISTS Courses(code TEXT, name TEXT,
                      slots TEXT, type TEXT, dt TEXT)""")
    cursor.execute("SELECT * FROM Schedules")
    if cursor.fetchall()==[]:
        cursor.executemany("INSERT INTO Schedules VALUES (?,?)",[("monday","[]"),
                                                             ("tuesday","[]"),
                                                             ("wednesday","[]"),
                                                             ("thursday","[]"),
                                                             ("friday","[]")])
    mycon.commit()


class day:
    def __init__(self, day, schedule = []):
        self.day = day
        self.schedule = schedule

class course:
    def __init__(self, code="", name="", slots=[], type="L", dt=[]):
        self.code = code
        self.name = name
        self.slots = slots
        self.type = type
        self.dt = dt


help = dict(help_option_names = ['-h','--help'])


class CustomGroup(click.Group):
    def format_commands(self, ctx, formatter):
        commands = []
        for subcommand in self.list_commands(ctx):
            cmd = self.get_command(ctx, subcommand)
            # What is this, the tool lied about a command.  Ignore it
            if cmd is None:
                continue
            if cmd.hidden:
                continue

            commands.append((subcommand, cmd))

        # allow for 3 times the default spacing
        if len(commands):
            limit = formatter.width - 6 - max(len(cmd[0]) for cmd in commands)

            rows = []
            for subcommand, cmd in commands:
                help = cmd.get_short_help_str(limit)
                rows.append((subcommand, help))

            if rows:
                with formatter.section('COMMANDS'):
                    formatter.write_dl(rows)

    def format_options(self, ctx, formatter):
        """Writes all the options into the formatter if they exist."""
        opts = []
        for param in self.get_params(ctx):
            rv = param.get_help_record(ctx)
            if rv is not None:
                if rv[0] == '-h, --help':
                    rv = list(rv)
                    rv[1] = "Show this message."
                    rv = tuple(rv)
                opts.append(rv)
        if opts:
            with formatter.section('OPTIONS'):
                formatter.write_dl(opts)

    def format_help(self, ctx, formatter):
        head = """__      __ _____  _______  _                  \n\\ \\    / /|_   _||__   __|(_)                 \n \\ \\  / /   | |     | |    _  _ __ ___    ___ \n  \\ \\/ /    | |     | |   | || '_ ` _ \\  / _ \\\n   \\  /    _| |_    | |   | || | | | | ||  __/\n    \\/    |_____|   |_|   |_||_| |_| |_| \\___|\n\n"""
        body = """CLI tool to view your timetable from terminal anytime!

Developer :
  Dhruv Shah (https://github.com/Dhruv9449)\n"""

        formatter.write(head)
        formatter.write(body)
        formatter.write("\nFor more detailed help:\n  https://github.com/Dhruv9449/VITime-CLI/blob/main/README.md\n")
        self.format_commands(ctx, formatter)
        formatter.write("\nUSAGE:\n"+"  vitime [COMMAND] [COMMAND OPTIONS]\n")
        self.format_options(ctx, formatter)


class CustomCommand(click.Command):
    def format_options(self, ctx, formatter):
        """Writes all the options into the formatter if they exist."""
        opts = []
        for param in self.get_params(ctx):
            rv = param.get_help_record(ctx)
            if rv is not None:
                if rv[0] == '-h, --help':
                    rv = list(rv)
                    rv[1] = "Show this message."
                    rv = tuple(rv)
                elif rv[0] == '-n, --name TEXT':
                    rv = list(rv)
                    rv[0] = "-n, --name     "
                    rv = tuple(rv)
                opts.append(rv)
        if opts:
            with formatter.section('OPTIONS'):
                formatter.write_dl(opts)
        return opts

    def format_help(self, ctx, formatter):
        print("             VITIME\n\nCOMMAND:")
        path = ctx.command_path
        print("  "+ctx.command_path+"\n")
        print("DESCRIPTION:")
        text = self.help
        print("  "+text)
        self.format_options(ctx, formatter)
        print("\nUSAGE:\n  "+path+" [OPTIONS]\n")
        print("For more detailed help:\n  https://github.com/Dhruv9449/VITime-CLI/blob/main/README.md\n")



#Database functions
def addcourse_db(Course):
    cursor.execute("INSERT INTO Courses values (?,?,?,?,?)",
                    (Course.code, Course.name, str(Course.slots), Course.type, str(Course.dt)))
    mycon.commit()

def addschedule_db(Day):
    cursor.execute("UPDATE Schedules SET schedule = ? WHERE day = ?", (str(Day.schedule), Day.day,))
    mycon.commit()

def loadcourse(code):
    cursor.execute("SELECT * FROM Courses WHERE code = ?",(code,))
    code, name, slots, type, dt = cursor.fetchall()[0]
    slots, dt = eval(slots), eval(dt)
    Course = course(code, name, slots, type, dt)
    return Course

def loadday(name):
    cursor.execute("SELECT * FROM Schedules WHERE day = ?",(name,))
    name, schedule = cursor.fetchall()[0]
    schedule = eval(schedule)
    Day = day(name, schedule)
    return Day

def checkcourse(Name):
    if not re.search("[A-Z][A-Z][A-Z][A-Z]\d\d\d|[A-Z][A-Z][A-Z]\d\d\d\d", Name.code):
        print(f"\n{Name.code} is an invalid course code, please enter a valid course code.(eg - BCHY101L/CSE1002)\n")
        return True
    cursor.execute("SELECT * from Courses WHERE code = ? AND slots = ?",(Name.code, str(Name.slots)))
    courses = cursor.fetchall()
    if len(courses) == 0:
        return False
    print(f"\nCourse {Name.code} already exists!\n")
    return True


def checktype(type):
    if type == "L" :
        return "Theory"
    else:
        return "Lab"

#input command
def retrieveinput():
    print("\nGo to vtop and copy all the courses.")
    print("For more help on how to do that check -\n  https://github.com/Dhruv9449/VITime-CLI#vitime-addtimetable")
    confirm = input("\nType 'yes' if you have copied all the courses : ").lower()
    if confirm == 'yes' :
        text = pyperclip.paste().replace('\r','\n')
        return text
    else:
        print("\nPlease go copy all your courses and try again!")
        return False


def scrape(data):
    try:
        code_name = [i[:-1] for i in re.findall("[A-Z][A-Z][A-Z][A-Z]\d\d\d.+\n|[A-Z][A-Z][A-Z]\d\d\d\d.+\n", data)]
        code = [i.split()[0] for i in code_name]
        name = [i.split(" - ")[1] for i in code_name]
        slots = [j[:-3].split('+') for j in re.findall(".+\d.+[-]\n|NIL.+[-]\n",data)]
        return list(zip(code, name, slots))

    except IndexError :
        return []


VITimeCLI/Slotdata.py
daysTh = {"monday" : ['A1', 'F1', 'D1', 'TB1', 'TG1', 'A2', 'F2', 'D2', 'TB2', 'TG2', 'V3'],
          "tuesday" : ['B1', 'G1', 'E1', 'TC1', 'TAA1', 'B2', 'G2', 'E2', 'TC2', 'TAA2', 'V4'],
          "wednesday" : ['C1', 'A1', 'F1', 'V1', 'V2', 'C2', 'A2', 'F2', 'TD2', 'TBB2', 'V5'],
          "thursday" : ['D1', 'B1', 'G1', 'TE1', 'TCC1', 'D2', 'B2', 'G2', 'TE2', 'TCC2', 'V6'],
          "friday" : ['E1', 'C1', 'TA1', 'TF1', 'TD1', 'E2', 'C2', 'TA2', 'TF2', 'TDD2', 'V7']}

thSlots = ["08:00 - 08:50" ,
           "09:00 - 09:50" ,
           "10:00 - 10:50" ,
           "11:00 - 11:50" ,
           "12:00 - 12:50" ,
           "14:00 - 14:50" ,
           "15:00 - 15:50" ,
           "16:00 - 16:50" ,
           "17:00 - 17:50" ,
           "18:00 - 18:50" ,
           "19:00 - 19:50" ]

daysLb =  {"monday" : ['L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'L31', 'L32', 'L33', 'L34', 'L35', 'L36'],
           "tuesday" : ['L7', 'L8', 'L9', 'L10', 'L11', 'L12', 'L37', 'L38', 'L39', 'L40', 'L41', 'L42'],
           "wednesday" : ['L13', 'L14', 'L15', 'L16', 'L17', 'L18', 'L43', 'L44', 'L45', 'L46', 'L47', 'L48'],
           "thursday" : ['L19', 'L20', 'L21', 'L22', 'L23', 'L24', 'L49', 'L50', 'L51', 'L52', 'L53', 'L54'],
           "friday" : ['L25', 'L26', 'L27', 'L28', 'L29', 'L30', 'L55', 'L56', 'L57', 'L58', 'L59', 'L60']}

lbSlots = ['08:00 - 08:50',
           '08:51 - 09:40',
           '09:51 - 10:40',
           '10:41 - 11:30',
           '11:40 - 12:30',
           '12:31 - 13:20',
           '14:00 - 14:50',
           '14:51 - 15:40',
           '15:51 - 16:40',
           '16:41 - 17:30',
           '17:40 - 18:30',
           '18:31 - 19:20']

days = ["monday",
        "tuesday",
        "wednesday",
        "thursday",
        "friday"]

theoryslots = ['A1', 'F1', 'D1', 'TB1', 'TG1', 'A2', 'F2', 'D2', 'TB2', 'TG2', 'V3',
                'B1', 'G1', 'E1', 'TC1', 'TAA1', 'B2', 'G2', 'E2', 'TC2', 'TAA2', 'V4',
                'C1', 'A1', 'F1', 'V1', 'V2', 'C2', 'A2', 'F2', 'TD2', 'TBB2', 'V5',
                'D1', 'B1', 'G1', 'TE1', 'TCC1', 'D2', 'B2', 'G2', 'TE2', 'TCC2', 'V6',
                 'E1', 'C1', 'TA1', 'TF1', 'TD1', 'E2', 'C2', 'TA2', 'TF2', 'TDD2', 'V7']

labslots = ['L1', 'L2', 'L3', 'L4', 'L5', 'L6', 'L31', 'L32', 'L33', 'L34', 'L35', 'L36',
            'L7', 'L8', 'L9', 'L10', 'L11', 'L12', 'L37', 'L38', 'L39', 'L40', 'L41', 'L42',
            'L13', 'L14', 'L15', 'L16', 'L17', 'L18', 'L43', 'L44', 'L45', 'L46', 'L47', 'L48',
            'L19', 'L20', 'L21', 'L22', 'L23', 'L24', 'L49', 'L50', 'L51', 'L52', 'L53', 'L54',
            'L25', 'L26', 'L27', 'L28', 'L29', 'L30', 'L55', 'L56', 'L57', 'L58', 'L59', 'L60']
